File: chap2/julia/code1.jl
using CSV
using DataFrames
using Plots
using Random
using StatsBase

## オリジナルデータをUCI ML databaseから読み込む
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"
download(url, "wine.csv")
Wine = CSV.read("wine.csv", header=false)

## 使用する変数(品種, フラボノイド含有量, 色強度)を抽出
wine = Wine[:, [1, 8, 11]]
# 3変数の命名
names!(wine, [:grape, :x1, :x2])
wine[:ID] = 1:size(wine)[1]

## サンプルサイズの計算
# 品種ごと
n1 = count(grape -> (grape == 1), wine[:, 1])
n2 = count(grape -> (grape == 2), wine[:, 1])
n3 = count(grape -> (grape == 3), wine[:, 1])
# トータル
n = n1 + n2 + n3

## 乱数を用いて各品種から教師サンプルを選択(教師サンプルサイズは表2.1の設定)
# 各品種の教師サンプルサイズ
n1Train = 29
n2Train = 36
n3Train = 24
# 乱数種を指定
Random.seed!(380)
# 教師サンプル(品種1)
train1 = sample(wine[wine.grape .== 1, :].ID, n1Train, replace=false, ordered=true)
# 教師サンプル(品種2)
train2 = sample(wine[wine.grape .== 2, :].ID, n2Train, replace=false, ordered=true)
# 教師サンプル(品種3)
train3 = sample(wine[wine.grape .== 3, :].ID, n3Train, replace=false, ordered=true)
# 教師サンプル(全品種)
train = vcat((train1, train2, train3)...)

## 残りをテストサンプルとする
n1Test = n1 - n1Train
n2Test = n2 - n2Train
n3Test = n3 - n3Train
# テストサンプル(全品種)
test = setdiff(Vector(1:n), train)

## 教師データ・テストデータへ分割
wineTrain = wine[wine.ID .∈ Ref(train), :]
wineTest = wine[wine.ID .∈ Ref(test), :]

## 横軸:x1, 縦軸:x2 で散布図作成(図2.1に対応)
# x1, x2 の値域
r1 = (minimum(wine[:, 2]), maximum(wine[:, 2]))
r2 = (minimum(wine[:, 3]), maximum(wine[:, 3]))
# プロット画面を２分割
plts = Vector([])

# 教師データの散布図
plt = plot(
    title = "training; n1 = $n1Train, n2 = $n2Train, n3 = $n3Train",
    legend=:topright
)
xlims!(r1)
ylims!(r2)
xlabel!("x1")
ylabel!("x2")
for i in 1:3
    scatter!(
        wineTrain[wineTrain.grape .== i, :].x1,
        wineTrain[wineTrain.grape .== i, :].x2,
        markercolor = i,
        label = "grape$i"
    )
end
push!(plts, plt)

# 教師データの散布図
plt = plot(
    title = "test; n1 = $n1Test, n2 = $n2Test, n3 = $n3Test",
    legend=:topright
)
xlims!(r1)
ylims!(r2)
xlabel!("x1")
ylabel!("x2")
for i in 1:3
    scatter!(
        wineTest[wineTest.grape .== i, :].x1,
        wineTest[wineTest.grape .== i, :].x2,
        markercolor = i,
        label = "grape$i"
    )
end
push!(plts, plt)

## 判別境界を作図する際に使用する格子点を作成
_grid = Matrix(DataFrame(vec(collect(Iterators.product(r1[1]:(r1[2]-r1[1])/499:r1[2], r2[1]:(r2[2]-r2[1])/499:r2[2])))))

File: chap2/R/code1.r
## オリジナルデータをUCI ML databaseから読み込む
url <- "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"
Wine <- read.csv(url, header=FALSE)

## 使用する変数(品種, フラボノイド含有量, 色強度)を抽出
wine = Wine[, c(1, 8, 11)]
colnames(wine) = c("grape", "x1", "x2") # 3変数の命名

## サンプルサイズの計算
# 品種ごと
n1 = sum(wine[, 1] == 1)
n2 = sum(wine[, 1] == 2)
n3 = sum(wine[, 1] == 3)
# トータル
n = n1 + n2 + n3

## 乱数を用いて各品種から教師サンプルを選択(教師サンプルサイズは表2.1の設定)
n1.train = 29; n2.train = 36; n3.train = 24; # 各品種の教師サンプルサイズ
set.seed(380) # 乱数種を指定
# 教師サンプル(品種1)
train1 = sort(sample(which(wine[, 1] == 1), n1.train))
# 教師サンプル(品種2)
train2 = sort(sample(which(wine[, 1] == 2), n2.train))
# 教師サンプル(品種3)
train3 = sort(sample(which(wine[, 1] == 3), n3.train))
# 教師サンプル(全品種)
train = c(train1, train2, train3)

## 残りをテストサンプルとする
n1.test = n1 - n1.train
n2.test = n2 - n2.train
n3.test = n3 - n3.train
# テストサンプル(全品種)
test = (1:n)[-train]

## 教師データ・テストデータへ分割
wine.train = wine[train, ]; wine.test = wine[test, ]

## 横軸:x1, 縦軸:x2 で散布図作成(図2.1に対応)
r1 = range(wine[, 2]); r2 = range(wine[, 3]) # x1, x2 の値域
par(mfrow = c(1, 2)) # プロット画面を２分割

# 教師データの散布図
plot(
    wine.train[, 2], 
    wine.train[, 3], 
    xlim = r1, 
    ylim = r2,
    xlab = "x1", 
    ylab = "x2", 
    col = wine.train[, 1], 
    pch = wine.train[, 1], 
    main = paste0(
        "training; n1 = ", n1.train, ", n2 = ", n2.train, ", n3 =", n3.train
    )
)
legend(
    "topright", 
    legend = paste0(
        "grape", 
        1:3
    ), 
    col = 1:3, 
    pch = 1:3
)

# テストデータの散布図
plot(
    wine.test[, 2], 
    wine.test[, 3], 
    xlim = r1, 
    ylim = r2,
    xlab = "x1", 
    ylab = "x2", 
    col = wine.test[, 1], 
    pch = wine.test[, 1], 
    main = paste0(
        "test; n1 = ", n1.train, ", n2 = ", n2.train, ", n3", n3.train)
    )
legend(
    "topright", 
    legend = paste0(
        "grape", 1:3
    ), 
    col = 1:3, 
    pch = 1:3
)

par(mfrow = c(1, 1)) # プロット画面の分割を解除

## 判別境界を作図する際に使用する格子点を作成
grid = data.frame(
    expand.grid(
        x1 = seq(
            r1[1], 
            r1[2], 
            length.out = 500
        ), 
        x2 = seq(
            r2[1], 
            r2[2], 
            length.out = 500
        )
    )
)
